sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BeamFnDataReadRunner.java

```diff
@@ -166,14 +177,32 @@ public void registerInputLocation() {
             apiServiceDescriptor,
             LogicalEndpoint.of(processBundleInstructionIdSupplier.get(), inputTarget),
             coder,
-            receiver);
+            enqueueElementReceiver);
   }
 
   public void blockTillReadFinishes() throws Exception {
     LOG.debug(
         "Waiting for process bundle instruction {} and target {} to close.",
         processBundleInstructionIdSupplier.get(),
         inputTarget);
-    readFuture.awaitCompletion();
+
+    boolean draining = false;
+    while (true) {
+      WindowedValue<OutputT> val = this.queue.poll(50, TimeUnit.MILLISECONDS);
+      if (val != null) {
+        // Call receiver.
+        elementReceiver.accept(val);
+      } else if (draining) {
+        break;
+      }
+      if (readFuture.isDone()) {
```

Comment:
Thinking about this some more, this can't happen since readFuture can only be completed if the last element has already been transferred through the synchronous queue which implies the queue had someone "receiving" it.

Reply:
Agreed, rm'd